// appium-ci/build/lib/ios-tools.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.spawnAsUser = spawnAsUser;
exports.spawnAsCurrentUser = spawnAsCurrentUser;
exports.setIosSimulatorScale = setIosSimulatorScale;
exports.configureXcode = configureXcode;
exports.resetSims = resetSims;
exports.killAll = killAll;

var _path = _interopRequireDefault(require("path"));

var _utils = _interopRequireDefault(require("./utils"));

var _lodash = _interopRequireDefault(require("lodash"));

var _bluebird = _interopRequireDefault(require("bluebird"));

var _appiumSupport = require("appium-support");

const log = _appiumSupport.logger.getLogger('ios-tools');

const SIDE_DISK = '/Volumes/SIDE';

const SIDE_SIMS = _path.default.resolve(SIDE_DISK, 'sims');

function killAll(processes = ['instruments', 'simulator']) {
  processes = _lodash.default.flatten([processes]);
  const seq = (0, _lodash.default)(processes).map(p => {
    return () => {
      return _utils.default.exec(`sudo pkill -f ${p}`).catch(() => {});
    };
  }).value();
  return _bluebird.default.reduce(seq, function (_, fn) {
    return fn();
  }, null);
}

function spawnAsUser(user, cmd, args = []) {
  log.info(`Running spawnAsUser: ${user} ${cmd} ${args}`);
  return _utils.default.exec(`ps -axj | grep loginwindow | awk "/^${user} / {print \\$2;exit}"`).spread(function (stdout) {
    const userPid = stdout.trim();
    return _utils.default.spawn('sudo', ['launchctl', 'bsexec', userPid, 'sudo', '-u', user, cmd, ...args], {
      detached: false
    });
  });
}

function spawnAsCurrentUser(cmd, args = []) {
  log.info(`Running spawnAsCurrentUser: ${cmd} ${args}`);
  return _utils.default.exec('whoami').spread(function (stdout) {
    const currentUser = stdout.trim();
    return spawnAsUser(currentUser, cmd, args);
  });
}

function setIosSimulatorScale(scale = '0.5') {
  log.info('Setting simulator scale');
  return _utils.default.exec(`defaults write com.apple.iphonesimulator SimulatorWindowLastScale ${scale}`);
}

function configureXcode(xCodeVersion) {
  log.info(`Configuring xCode: ${xCodeVersion}`);

  const bin = _path.default.resolve(SIDE_SIMS, 'configure.sh');

  return _utils.default.exec(`${bin} ${xCodeVersion}`);
}

function resetSims() {
  log.info('Resetting simulators');

  const bin = _path.default.resolve(SIDE_SIMS, 'reset-sims.sh');

  return _utils.default.exec(bin);
}require('source-map-support').install();
