// appium-mac-driver/build/lib/driver.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.MacDriver = void 0;

require("source-map-support/register");

var _appiumBaseDriver = require("appium-base-driver");

var _appiumSupport = require("appium-support");

var _appiumForMac = require("./appium-for-mac");

var _logger = _interopRequireDefault(require("./logger"));

const NO_PROXY_LIST = [['POST', /execute/]];

class MacDriver extends _appiumBaseDriver.BaseDriver {
  constructor(opts = {}, shouldValidateCaps = true) {
    super(opts, shouldValidateCaps);
    this.jwpProxyActive = false;
    this.opts.address = opts.address || _appiumForMac.DEFAULT_A4M_HOST;
  }

  async createSession(...args) {
    if (!_appiumSupport.system.isMac()) {
      throw new Error('AppiumForMac tests only run on the Mac');
    }

    try {
      let [sessionId, caps] = await super.createSession(...args);
      await this.startAppiumForMacSession();

      if (caps.app) {
        _logger.default.info(`Automatically navigating to app '${caps.app}'`);

        await this.a4mDriver.sendCommand('/url', 'POST', {
          url: caps.app
        });
      }

      return [sessionId, caps];
    } catch (e) {
      await this.deleteSession();
      throw e;
    }
  }

  async startAppiumForMacSession() {
    this.a4mDriver = new _appiumForMac.AppiumForMac();
    await this.a4mDriver.start();
    await this.a4mDriver.startSession(this.caps);
    this.proxyReqRes = this.a4mDriver.proxyReqRes.bind(this.a4mDriver);
    this.jwpProxyActive = true;
  }

  async deleteSession() {
    _logger.default.debug('Deleting AppiumForMac session');

    if (this.a4mDriver && this.jwpProxyActive) {
      await this.a4mDriver.deleteSession();
      await this.a4mDriver.stop();
      this.a4mDriver = null;
    }

    this.jwpProxyActive = false;
    await super.deleteSession();
  }

  async execute(script, args) {
    if (!this.relaxedSecurityEnabled) {
      _logger.default.errorAndThrow(`Appium server must have relaxed security flag set in order to run any shell commands`);
    }

    return await this.a4mDriver.sendCommand(`/session/${this.sessionId}/execute`, 'POST', {
      script,
      args
    });
  }

  proxyActive() {
    return true;
  }

  getProxyAvoidList() {
    return NO_PROXY_LIST;
  }

  canProxy() {
    return true;
  }

  get driverData() {
    return {
      A4MPort: this.opts.port
    };
  }

}

exports.MacDriver = MacDriver;
var _default = MacDriver;
exports.default = _default;require('source-map-support').install();
